### DISCUSSION TITLE
apt 安装应用，dde-lanchpad 没有图标

### DISCUSSION THREAD

**Discussion by tensor-tech** [linuxdeepin/developer-center#7384] [Developer Discussion | 开发者讨论]
目前系统版本：deepin23b3
启动器版本：0.4.6
我进行的操作：sudo apt purge deepin-app-store,sudo  apt install com.sublimetext.sublime-text-4 安装成功后，启动器没有sublime图标

**Comment by BLumia** [comment]
> `sudo apt install com.sublimetext.sublime-text-4` 安装成功后，启动器没有sublime图标

没有图标指的是图标不显示（显示为空白）还是说启动器里整个压根找不到这个应用？
如果是压根搜不到的话，dpkg -L 下这个包，desktop 文件是放在哪里了？如果 `killall dde-launchpad` 后重新点出启动器，会有图标吗？

**Reply by tensor-tech** [reply]
感谢回复！
我的操作：sudo apt purge deepin-app-store, 重启电脑，sudo apt install com.sublimetext.sublime-text-4
1. 启动器压根找不到
2. 在/opt/apps/com.sublimetext.sublime-text-4 能找到这个应用，并且通过的desktop文件能够启动。
3. dpkg -L 的结果：
[dpkg-L.txt](https://github.com/linuxdeepin/developer-center/files/14504524/dpkg-L.txt)


4.killall dde-launchpad 后重新点出启动器，不会有图标

**Reply by BLumia** [reply]
卸载深度商店应该与launchpad找不到这个应用的问题不相关，因为 beta3 所提供的 launchpad 版本走的是 gio 获取的应用列表而不是 deepin 的 dde-application-manager。

不过如果你很明确是卸载商店才会导致这个问题的话，可以对比下不卸载商店时与卸载并重启后 `env | grep XDG_DATA` 的输出的差异（主要关注 `XDG_DATA_DIRS` 这个环境变量，这个决定了 gio 会去哪些地方找应用程序的 desktop 文件。

> `/opt/apps/com.sublimetext.sublime-text-4/entries/applications/com.sublimetext.sublime-text-4.desktop` （截自你的 dpkg -L 输出）

比如这个路径应该不会被 gio 默认扫描进去吧？

顺便，后续 launchpad 会改由 dde-application-manager 获取应用列表，不过 dde-application-manager 本身也会遵循 `XDG_DATA_DIRS` 来获取应用列表。

**Reply by tensor-tech** [reply]
感谢回复！
首先确定是我卸载商店引起的这个问题，因为我重装回商店，再去通过apt 安装 sublime，launchpad 就会有图标。
根据您的提示，卸载之后，并且重启电脑，XDG_DATA_DIRS 输出和卸载之前是完全一致的, 确实没有/opt/apps,如下：
```
XDG_DATA_HOME=/home/ffff/.local/share
XDG_DATA_DIRS=/home/ffff/.local/share/linglong/entries/share:/var/lib/linglong/entries/share:/usr/local/share:/usr/share
```
但是我发现应该不是它的问题，我重新安装了商店，然后通过apt reinstall com.sublimetext.sublime-text-4 重新安装sublime，launchpad 就会有图标。
我对比了一下apt  安装 sublime的输出，安装了商店会触发一个deepin-app-store的触发器，全部输出如下，大概就是这个触发器。
```
ffff@ffff-PC:~/Desktop$ sudo apt reinstall  com.sublimetext.sublime-text-4 
正在读取软件包列表... 完成
正在分析软件包的依赖关系树... 完成
正在读取状态信息... 完成                 
下列软件包是自动安装的并且现在不需要了：
  boot-info-script jfsutils libutempter0 smartmontools squashfs-tools x11-apps x11-session-utils xbitmaps
  xinit
使用'sudo apt autoremove'来卸载它(它们)。
升级了 0 个软件包，新安装了 0 个软件包，重新安装了 1 个软件包，要卸载 0 个软件包，有 2 个软件包未被升级。
需要下载 15.9 MB 的归档。
解压缩后会消耗 0 B 的额外空间。
获取:1 https://com-store-packages.uniontech.com/appstorev23 beige/appstore amd64 com.sublimetext.sublime-text-4 amd64 4169 [15.9 MB]
已下载 15.9 MB，耗时 4秒 (4,525 kB/s)                      
(正在读取数据库 ... 系统当前共安装有 257913 个文件和目录。)
准备解压 .../com.sublimetext.sublime-text-4_4169_amd64.deb  ...
正在解压 com.sublimetext.sublime-text-4 (4169) 并覆盖 (4169) ...
正在设置 com.sublimetext.sublime-text-4 (4169) ...
正在处理用于 libc-bin (2.37-12deepin3) 的触发器 ...
正在处理用于 deepin-app-store (8.0.6.0011-1) 的触发器 ...
Rebuilding /usr/share/applications/bamf-2.index...
```

**正在处理用于 deepin-app-store (8.0.6.0011-1) 的触发器 ...**
如果卸载了商店是没有这个触发器的。

**Reply by tensor-tech** [reply]
补充：卸载商店之后，apt 安装的输出：
ffff@ffff-PC:~/Desktop$ sudo apt reinstall com.sublimetext.sublime-text-4 
正在读取软件包列表... 完成
正在分析软件包的依赖关系树... 完成
正在读取状态信息... 完成                 
下列软件包是自动安装的并且现在不需要了：
  boot-info-script jfsutils libutempter0 smartmontools squashfs-tools x11-apps x11-session-utils xbitmaps
  xinit
使用'sudo apt autoremove'来卸载它(它们)。
升级了 0 个软件包，新安装了 0 个软件包，重新安装了 1 个软件包，要卸载 0 个软件包，有 2 个软件包未被升级。
需要下载 15.9 MB 的归档。
解压缩后会消耗 0 B 的额外空间。
获取:1 https://com-store-packages.uniontech.com/appstorev23 beige/appstore amd64 com.sublimetext.sublime-text-4 amd64 4169 [15.9 MB]
已下载 15.9 MB，耗时 11秒 (1,406 kB/s)                                                                       
(正在读取数据库 ... 系统当前共安装有 257566 个文件和目录。)
准备解压 .../com.sublimetext.sublime-text-4_4169_amd64.deb  ...
正在解压 com.sublimetext.sublime-text-4 (4169) 并覆盖 (4169) ...
正在设置 com.sublimetext.sublime-text-4 (4169) ...
正在处理用于 libc-bin (2.37-12deepin3) 的触发器 ...

**Reply by BLumia** [reply]
那这样的话，感觉商店可能以别的形式干预了 `XDG_DATA_DIRS` （没细看，不太确定）。

主干分支的启动器已经改为使用 dde-application-manager 获取应用程序列表了，那这样的话你可以使用 d-feet 或者 qdbusviewer 之类的程序检查下 dde-application-manager 本身所提供的 dbus 服务下有没有 sublime 对应的项目。如果 dde-application-manager 一直都是正常的话，也不必担心启动器后续找不到这个应用的问题了。如果 dde-application-manager 不正常的话可能需要进一步确认问题了。

ps. 你在开发者 matrix/tg 群吗？如果没的话可以考虑加一下 :)

**Reply by BLumia** [reply]
update: 在开发者群确认后，application-manager 获取的信息正常，不再需要额外修改。
